<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

In [1]:
import pandas as pd
import numpy as np


from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split,cross_val_score


In [2]:
df= pd.read_csv('/datasets/insurance.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [4]:
df.columns

Index(['Пол', 'Возраст', 'Зарплата', 'Члены семьи', 'Страховые выплаты'], dtype='object')

In [5]:
df.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

In [6]:
df.head(10)

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0
5    1     43.0   41000.0            2                  1
6    1     39.0   39700.0            2                  0
7    1     25.0   38600.0            4                  0
8    1     36.0   49700.0            1                  0
9    1     32.0   51700.0            1                  0

- Пропусков в данных нет;
- Переименуем столбцы согласно правилам;
- Стоит изменить столбцы возвраста и зарплаты на целочисленные.

In [7]:
df = df.rename(columns={'Пол':'gender','Возраст':'age',
                      'Зарплата':'salary','Члены семьи':'family',
                      'Страховые выплаты':'payment'})

In [8]:
df[['age','salary']] = df[['age','salary']].astype('int64')

In [9]:
df.head(10)

gender  age  salary  family  payment
0       1   41   49600       1        0
1       0   46   38000       1        1
2       0   29   21000       0        0
3       0   21   41700       2        0
4       1   28   26100       0        0
5       1   43   41000       2        1
6       1   39   39700       2        0
7       1   25   38600       4        0
8       1   36   49700       1        0
9       1   32   51700       1        0

Дальнейших обработок данных не требуется,можно приступать к следующему шагу.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии?

Для предсказания будет использоваться матрица признаков 'gender','age','salary','family',серия с выплатами будет целевым признаком

Формула обучения линейной регресии:

$$
w = (X^T X)^{-1} X^T y
$$

Новая матрица будет иметь вид 
 
 $$A=XP$$
 
Новые предсказания будут иметь вид
 
$$a_n =  Aw_n = XPw_n$$

Новая формула обучения 

$$w_n=((XP^T) XP))^{-1} (XP)^T y $$

Согласно свойствам транспонирования матриц $(XP)^T=P^TX^T$,перепишем уравнение

$$w_n = (P^TX^TXP)^{-1}P^TX^Ty$$

Так же матрицы  $X$  и $X^T$ прямоугольные, поэтому для них не существует обратных матриц и  нельзя писать раздельно, поэтому значение в скобках $((XP^T) XP))^{-1}$ мы оставим как есть, а $P$ явлется квадратной матрицей. При раскрытии скобок нужно менять порядок множителей местами, поэтому $P$ ставим в начало

$$w_n = P^{-1}(X^TX)^{-1}(P^T)^{-1} P^TX^Ty$$

Произведение $(P^T)^{-1}$ и $P^T$ представляет собой единичную матрицу,сократим их

$$w_n=P^{-1}(X^TX)^{-1}X^Ty$$

Поскольку $(X^T X)^{-1} X^T y$ является мерой весов в модели или $w$ - перепишем уравнение

$$w_n=P^{-1}w$$

Тогда новые предсказания будут иметь вид
$$a_n = XPP^{-1}w$$

Исходя из правила, что умножение квадратной матрицы $А$ на обратную ей $А^{-1}$ получается так же единичная матрица:

$$a_n = Xw = a$$

**Таким образом, качество регрессии не изменится,поскольку зашифрованные признаки будут фактически либо равны либо крайне близки по метрике оригиналу**

**Ответ:** *При шифровании признаков качество регрессии не изменится*

**Обоснование:** *Это обусловлено тем, что при формуле обучения линейной регресии новые зашифрованные признаки будут равны или кране близки к старым*

## Алгоритм преобразования

**Алгоритм**

1) Первым делом мы разделим признаки на фичи и целевой признак;

2) Возьмем случайную квадратную матрицу;

3) Для начала обучим модель линейной регрессии на обычных,незашифрованных данных;

4) Посчитаем для этих данных метрику r2_score;

5) Далее фичи умножим на нашу квадратную случайно сгенерированную матрицу;

6) У нее так же посчитаем метрику r2_score;

7) Далее сравним, какая получилась разница между этими двумя метриками

9) Сделаем выводы по полученным результатам.


**Обоснование**
Мы проведем проверку по коэффициенту детерминаии для двух моделей,обученных на зашифрованных и незашифрованных данных,далее сравним показатели. Так же посмотрим признаки модели, чтобы была возможность отследить, что могло измениться в наших моделях. Если же метрики сойдутся, идея будет доказана.

## Проверка алгоритма

Разделим выборки

In [11]:
features = df.drop('payment',axis=1)
target = df['payment']

features_train,features_test,target_train,target_test = train_test_split(features,target,test_size=0.33, random_state=12345)

Посмтрим размер выборок

In [12]:
print(features_train.shape)
print(features_test.shape)

(3350, 4)
(1650, 4)


In [13]:
print(target_train.shape)
print(target_test.shape)

(3350,)
(1650,)


Всё отлично, теперь обучим модель без шифрования и посомтрим,какой коэффициент детерминации у нас получится, а так же оценим параметры обсчёта модели, которые в дальнейшем мы сможем сравнить с моделью с измененными фичами.

In [14]:
model = LinearRegression()
model.fit(features_train,target_train)
predictions = model.predict(features_test)
r2 = r2_score(target_test,predictions)
              
print(f'Средний коэфициент детерминации R2 на тестовой выборке равен {r2}')
print(f"Коэффициенты получившейся модели {model.coef_}")
print(f"Intercept получившейся модели {model.intercept_}")

Средний коэфициент детерминации R2 на тестовой выборке равен 0.4269864803108606
Коэффициенты получившейся модели [ 1.46559127e-02  3.60925527e-02 -1.18312871e-07 -1.18313026e-02]
Intercept получившейся модели -0.9576320116572752


Далее напишем функцию для "шифровки наших параметров" и проверим ее так же на модели линейной регрессии

In [55]:
m = []

def m_check (m):
    m_new = np.linalg.inv(m)
    if np.allclose(np.dot(m, m_new), np.eye(m.shape[0])) is True:
        return  True
    else:
        return False


def anonim (features_train,target_train):
    m = np.random.normal(size=(features.shape[1], features.shape[1]))
    if m_check(m) is True:
        m_train = np.dot(features_train.values,m)
        m_test = np.dot(features_test.values,m)
        print('Матрица обратимая,можно строить модель')
        return m_train,m_test
    else: 
        print('Всё сломалось')

Применим нашу функцию для шифровки

In [56]:
features_train_new,features_test_new = anonim(features_train,target_train)

Матрица обратимая,можно строить модель


In [57]:
model = LinearRegression()
model.fit(features_train_new,target_train)
predictions = model.predict(features_test_new)
r2 = r2_score(target_test,predictions)
              
print(f'Средний коэфициент детерминации R2 на тестовой выборке равен {r2}')
print(f"Коэффициенты получившейся модели {model.coef_}")
print(f"Intercept получившейся модели {model.intercept_}")

Средний коэфициент детерминации R2 на тестовой выборке равен 0.42698648031084885
Коэффициенты получившейся модели [ 0.03728525  0.0858241   0.00259828 -0.00842229]
Intercept получившейся модели -0.9576320116554928


## Выводы

- Как мы видим, значение метрики не изменилось - **0.42699** в обоих случаях, для зашифрованных и оригинальных признаков;
- Параметр $intercept$ у зашифрованной и оригинальной модели так же одинаковые - **-0.9576**;
- Коэффициенты же у моделей различаются, это было ожидаемо, ведь мы умножаем наши признаки на матрицу, что влечет за собой изменение коэффициентов. На качество модели это, тем не менее, не сказывается, что и требовалось доказать.